packages/gatsby-page-utils/src/ignore-path.ts

```diff
@@ -0,0 +1,38 @@
+import * as mm from "micromatch"
+
+interface IPathIgnoreOptions {
+  patterns: string | ReadonlyArray<string>
+  options?: mm.Options
+}
+
+export function ignorePath(
+  path: string,
+  ignore?: IPathIgnoreOptions | string | string[] | {} | null
+): boolean {
+  // Don't do anything if no ignore patterns
+  if (!ignore) return false
+  const settings: {
+    patterns: string | ReadonlyArray<string>
+    options: mm.Options
+  } = {
+    patterns: ``,
+    options: {},
+  }
+  if ((ignore as IPathIgnoreOptions).options !== undefined) {
+    settings.options = (ignore as IPathIgnoreOptions).options || {}
+  }
+  if ((ignore as IPathIgnoreOptions).patterns) {
+    settings.patterns = (ignore as IPathIgnoreOptions).patterns
+  } else {
+    // Allow shorthand ignore patterns ['pattern'] or 'pattern'
+    if (Array.isArray(ignore) && ignore.length > 0) {
+      settings.patterns = ignore
+    } else if (typeof ignore === `string`) {
+      settings.patterns = ignore
+    } else {
+      return false
+    }
+  }
```

Comment:
If you reorder the checks then I expect TS to be able to infer the proper type so you don't have to cast.

```suggestion
  if (typeof ignore === `string`) {
    settings.patterns = ignore
  } else if (Array.isArray(ignore)) {
    if (ignore.length > 0) {
      settings.patterns = ignore
    }
  } else if (ignore === null) {
    return false
  } else {
    // TS should be able to assert that `ignore` must be a `IPathIgnoreOptions` now
    if (!ignore.options && !ignore.patterns) {
      return false
    }
    if (ignore.options) {
      settings.options = ignore.options
    }
    if (ignore.patterns) {
      settings.patterns = ignore.patterns
    }
  }    
```